File: src/main/java/com/dyonovan/jatm/common/blocks/storage/BlockTank.java
package com.dyonovan.jatm.common.blocks.storage;

import com.dyonovan.jatm.JATM;
import com.dyonovan.jatm.collections.DummyState;
import com.dyonovan.jatm.common.blocks.BlockBakeable;
import com.dyonovan.jatm.common.tileentity.storage.TileDiamondTank;
import com.dyonovan.jatm.common.tileentity.storage.TileGoldTank;
import com.dyonovan.jatm.common.tileentity.storage.TileIronTank;
import com.dyonovan.jatm.helpers.GuiHelper;
import com.dyonovan.jatm.lib.Constants;
import net.minecraft.block.material.Material;
import net.minecraft.block.state.IBlockState;
import net.minecraft.client.multiplayer.WorldClient;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.BlockPos;
import net.minecraft.util.ChatComponentText;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumWorldBlockLayer;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;
import net.minecraftforge.fluids.FluidContainerRegistry;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.FluidTankInfo;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import java.util.Random;

public class BlockTank extends BlockBakeable {
    protected int buckets;
    public BlockTank(String name, int bucketCount) {
        super(Material.glass, name, null);
        setCreativeTab(JATM.tabJATM);
        setUnlocalizedName(Constants.MODID + ":" + name);
        setHardness(3.0F);
        buckets = bucketCount;
    }

    public TileEntity createNewTileEntity(World worldIn, int meta) {
        switch (buckets) {
            case 3 :
                return new TileDiamondTank();
            case 2 :
                return new TileGoldTank();
            case 1 :
            default:
            return new TileIronTank();
        }
    }
    @Override
    public int getLightValue (IBlockAccess world, BlockPos pos)
    {
        TileEntity tank = world.getTileEntity(pos);
        if (tank != null && tank instanceof TileIronTank)
            return ((TileIronTank) tank).getBrightness();
        return 0;
    }

    @Override
    public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, EntityPlayer player, EnumFacing side, float hitX, float hitY, float hitZ) {
        super.onBlockActivated(world, pos, state, player, side, hitX, hitY, hitZ);
        TileIronTank tank = (TileIronTank)world.getTileEntity(pos);

        ItemStack heldItem = player.getHeldItem();
        if(heldItem != null) {
            FluidStack fluid = FluidContainerRegistry.getFluidForFilledItem(heldItem);
            if(fluid != null) {
                int amount = tank.fill(EnumFacing.UP, fluid, false);
                if(amount == fluid.amount) {
                    tank.fill(EnumFacing.UP, fluid, true);
                    if (!player.capabilities.isCreativeMode)
                        player.inventory.setInventorySlotContents(player.inventory.currentItem, consumeItem(heldItem));
                    return true;
                }
                else
                    return false;
            } else if (FluidContainerRegistry.isBucket(heldItem)) {
                FluidTankInfo[] tanks = tank.getTankInfo(EnumFacing.UP);
                FluidStack fillFluid = tanks[0].fluid;// getFluid();
                ItemStack fillStack = FluidContainerRegistry.fillFluidContainer(fillFluid, heldItem);
                if (fillStack != null) {
                    tank.drain(EnumFacing.DOWN, FluidContainerRegistry.getFluidForFilledItem(fillStack).amount, true);
                    if (!player.capabilities.isCreativeMode) {
                        if (heldItem.stackSize == 1) {
                            player.inventory.setInventorySlotContents(player.inventory.currentItem, fillStack);
                        }
                        else {
                            player.inventory.setInventorySlotContents(player.inventory.currentItem, consumeItem(heldItem));
                            if (!player.inventory.addItemStackToInventory(fillStack)) {
                                player.dropPlayerItemWithRandomChoice(fillStack, false);
                            }
                        }
                    }
                    return true;
                }
            }
        }
        world.markBlockForUpdate(pos);
        if(world.isRemote)
            player.addChatComponentMessage(new ChatComponentText(tank.tank.getFluid() == null ? "Empty" : GuiHelper.GuiColor.YELLOW + tank.tank.getFluid().getLocalizedName() + GuiHelper.GuiColor.WHITE + " : " + tank.tank.getFluidAmount() + "/" + tank.tank.getCapacity()));
        return super.onBlockActivated(world, pos, state, player, side, hitX, hitY, hitZ);
    }

    public static ItemStack consumeItem (ItemStack stack)
    {
        if (stack.stackSize == 1)
        {
            if (stack.getItem().hasContainerItem())
                return stack.getItem().getContainerItem(stack);
            else
                return null;
        }
        else
        {
            stack.splitStack(1);

            return stack;
        }
    }

    @Override
    public void breakBlock(World worldIn, BlockPos pos, IBlockState state)
    {
        ItemStack stack = new ItemStack(this, 1);
        TileIronTank logic = (TileIronTank) worldIn.getTileEntity(pos);
        FluidStack liquid = logic.tank.getFluid();
        if (liquid != null)
        {
            NBTTagCompound tag = new NBTTagCompound();
            NBTTagCompound liquidTag = new NBTTagCompound();
            liquid.writeToNBT(liquidTag);
            tag.setTag("Fluid", liquidTag);
            stack.setTagCompound(tag);
        }
        dropTankBlock(worldIn, pos, stack);

        worldIn.setBlockToAir(pos);
        worldIn.markBlockForUpdate(pos);
        super.breakBlock(worldIn, pos, state);
    }

    @Override
    public Item getItemDropped(IBlockState state, Random rand, int fortune){
        return null;
    }

    protected void dropTankBlock (World world, BlockPos pos, ItemStack stack)
    {
        if (!world.isRemote && world.getGameRules().getGameRuleBooleanValue("doTileDrops"))
        {
            float f = 0.7F;
            double d0 = (double) (world.rand.nextFloat() * f) + (double) (1.0F - f) * 0.5D;
            double d1 = (double) (world.rand.nextFloat() * f) + (double) (1.0F - f) * 0.5D;
            double d2 = (double) (world.rand.nextFloat() * f) + (double) (1.0F - f) * 0.5D;
            EntityItem entityitem = new EntityItem(world, (double) pos.getX() + d0, (double) pos.getY() + d1, (double) pos.getZ() + d2, stack);
            entityitem.setPickupDelay(10);
            world.spawnEntityInWorld(entityitem);
        }
    }

    @Override
    public void onBlockPlacedBy(World worldIn, BlockPos pos, IBlockState state, EntityLivingBase placer, ItemStack stack)
    {
        super.onBlockPlacedBy(worldIn, pos, state, placer, stack);
        if (stack.hasTagCompound())
        {
            NBTTagCompound liquidTag = stack.getTagCompound().getCompoundTag("Fluid");
            if (liquidTag != null)
            {
                FluidStack liquid = FluidStack.loadFluidStackFromNBT(liquidTag);
                TileIronTank logic = (TileIronTank) worldIn.getTileEntity(pos);
                logic.tank.setFluid(liquid);
            }
        }
    }

    @Override
    public IBlockState getExtendedState(IBlockState state, IBlockAccess world, BlockPos pos) {
        return new DummyState(world, pos, this);
    }

    @Override
    public boolean isBlockNormalCube() {
        return false;
    }

    @Override
    public boolean isOpaqueCube() {
        return false;
    }

    @SideOnly(Side.CLIENT)
    public boolean isTranslucent() {
        return true;
    }

    public boolean isFullCube()
    {
        return false;
    }

    @SideOnly(Side.CLIENT)
    public EnumWorldBlockLayer getBlockLayer() {
        return EnumWorldBlockLayer.CUTOUT;
    }

    @Override
    public boolean canRenderInLayer(EnumWorldBlockLayer layer)
    {
        return layer == EnumWorldBlockLayer.CUTOUT || layer == EnumWorldBlockLayer.TRANSLUCENT;
    }

}

File: src/main/scala/com/dyonovan/neotech/common/blocks/storage/BlockTank.scala
package com.dyonovan.neotech.common.blocks.storage

import com.dyonovan.neotech.common.blocks.BaseBlock
import com.dyonovan.neotech.common.items.ItemWrench
import com.dyonovan.neotech.common.tiles.storage.TileTank
import com.teambr.bookshelf.loadables.ILoadActionProvider
import com.teambr.bookshelf.notification.{Notification, NotificationHelper}
import net.minecraft.block.material.Material
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.item.EntityItem
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.{Item, ItemStack}
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.math.BlockPos
import net.minecraft.util.{BlockRenderLayer, EnumBlockRenderType, EnumFacing, EnumHand}
import net.minecraft.world.{IBlockAccess, World}
import net.minecraftforge.client.event.ModelBakeEvent
import net.minecraftforge.fluids.{FluidContainerRegistry, FluidUtil, IFluidContainerItem, IFluidHandler}
import net.minecraftforge.fml.relauncher.{Side, SideOnly}

import scala.util.Random

/**
  * This file was created for NeoTech
  *
  * NeoTech is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Dyonovan
  * @since August 16, 2015
  */
class BlockTank(name: String, tier: Int) extends BaseBlock(Material.GLASS, name, classOf[TileTank]) with ILoadActionProvider {

    setHardness(2.0F)
    setBlockBounds(1F / 16F, 0F, 1F / 16F, 15F / 16F, 1F,  15F/ 16F)

    override def onBlockActivated(world: World, pos: BlockPos, state: IBlockState, player: EntityPlayer,
                                  hand: EnumHand, heldItem: ItemStack, side: EnumFacing, hitX: Float, hitY: Float, hitZ: Float) : Boolean = {

        val heldItem = player.getHeldItemMainhand
        val tank = world.getTileEntity(pos).asInstanceOf[TileTank]

        if(heldItem != null && tank.isInstanceOf[IFluidHandler]) {
            if(FluidUtil.interactWithTank(heldItem, player, tank, side.getOpposite))
                return true
        }

        //Wrench
        if (player.getHeldItemMainhand != null && player.getHeldItemMainhand.getItem.isInstanceOf[ItemWrench] && player.isSneaking) {
            if (breakTank(world, pos, state)) {
                world.setBlockToAir(pos)
                world.removeTileEntity(pos)
                world.setBlockState(pos, state, 3)
            }
        } else if (world.isRemote && tank != null) {
            var fluidName: String = ""
            var fluidAmount: String = ""
            if (tank.getCurrentFluid != null) {
                fluidName = tank.tank.getFluid.getLocalizedName
                fluidAmount = tank.tank.getFluid.amount.toString + " / " + tank.tank.getCapacity + " mb"
            } else {
                fluidName = "Empty"
                fluidAmount = "0 / " + tank.tank.getCapacity + " mb"
            }

            val item = new ItemStack(Item.getItemFromBlock(state.getBlock), 1)
            val tag = new NBTTagCompound
            tank.writeToNBT(tag)
            item.setTagCompound(tag)
            val notify = new Notification(item, fluidName, fluidAmount)
            NotificationHelper.addNotification(notify)
        }
        if(heldItem == null)
            return false
        FluidContainerRegistry.isFilledContainer(heldItem) || heldItem.getItem.isInstanceOf[IFluidContainerItem]
    }

    def consumeItem(stack: ItemStack): ItemStack = {
        if (stack.stackSize == 1) {
            if (FluidContainerRegistry.isContainer(stack))
                stack.getItem.getContainerItem(stack)
            else null
        } else {
            stack.splitStack(1)
            stack
        }
    }

    override def canHarvestBlock(world: IBlockAccess, pos: BlockPos, player: EntityPlayer): Boolean = {
        if (!player.capabilities.isCreativeMode) {
            breakTank(player.getEntityWorld, pos, world.getBlockState(pos))
        } else {
            super.breakBlock(player.getEntityWorld, pos, world.getBlockState(pos))
        }
        false
    }

    private def breakTank(world: World, pos: BlockPos, state: IBlockState): Boolean = {
        if (world.isRemote) return false
        world.getTileEntity(pos) match {
            case tile: TileTank =>
                val item = new ItemStack(Item.getItemFromBlock(state.getBlock), 1)
                val tag = new NBTTagCompound
                val tileTag = new NBTTagCompound
                tile.writeToNBT(tileTag)
                tag.setTag("Fluid", tileTag)
                item.setTagCompound(tag)
                if (tile.tank.getFluid != null) {
                    val r = tile.tank.getFluid.amount.toFloat / tile.tank.getCapacity
                    val res = 16 - (r * 16).toInt
                    item.setItemDamage(res)
                } else
                    item.setItemDamage(16)
                dropItem(world, item, pos) //Drop it
                return true
            case _ =>
        }
        false
    }

    override def onBlockPlacedBy(world: World, pos: BlockPos, state: IBlockState, placer: EntityLivingBase, stack:
    ItemStack): Unit = {
        if(stack.hasTagCompound && !world.isRemote) { //If there is a tag and is on the server
            world.getTileEntity(pos).readFromNBT(stack.getTagCompound.getCompoundTag("Fluid")) //Set the tag
            world.getTileEntity(pos).setPos(pos) //Set the saved tag to here
            world.setBlockState(pos, state, 3) //Mark for update to client
        }
    }

    override def getItemDropped(state: IBlockState, rand: java.util.Random, fortune: Int): Item = {
        null
    }

    override def createNewTileEntity(world: World, meta: Int): TileEntity = {
        new TileTank(tier)
    }

    def getName: String = name

    def getTier: Int = tier

    private def dropItem(world: World, stack: ItemStack, pos: BlockPos): Unit = {
        val random = new Random
        if (stack != null && stack.stackSize > 0) {
            val rx = random.nextFloat * 0.8F + 0.1F
            val ry = random.nextFloat * 0.8F + 0.1F
            val rz = random.nextFloat * 0.8F + 0.1F

            val itemEntity = new EntityItem(world,
                pos.getX + rx, pos.getY + ry, pos.getZ + rz,
                new ItemStack(stack.getItem, stack.stackSize, stack.getItemDamage))

            if (stack.hasTagCompound)
                itemEntity.getEntityItem.setTagCompound(stack.getTagCompound)

            val factor = 0.05F

            itemEntity.motionX = random.nextGaussian * factor
            itemEntity.motionY = random.nextGaussian * factor + 0.2F
            itemEntity.motionZ = random.nextGaussian * factor
            world.spawnEntityInWorld(itemEntity)

            stack.stackSize = 0
        }
    }

    override def getRenderType(state : IBlockState) : EnumBlockRenderType = EnumBlockRenderType.MODEL

    override def isOpaqueCube(state : IBlockState) : Boolean = false

    @SideOnly(Side.CLIENT)
    override def isTranslucent(state : IBlockState) : Boolean = true

    override def isFullCube(state : IBlockState) = false

    @SideOnly(Side.CLIENT)
    override def getBlockLayer : BlockRenderLayer = BlockRenderLayer.CUTOUT

    override def canRenderInLayer(layer: BlockRenderLayer): Boolean = {
        layer == BlockRenderLayer.CUTOUT || layer == BlockRenderLayer.TRANSLUCENT
    }

    override def performLoadAction(event: AnyRef, isClient: Boolean): Unit = {
        event match {
            case modelBake : ModelBakeEvent =>
             case _ =>
        }
    }
}
